// @terrestris/react-geo/dist/Field/ScaleCombo/ScaleCombo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

require("antd/es/select/style");

var _select = _interopRequireDefault(require("antd/es/select"));

var React = _interopRequireWildcard(require("react"));

var _Logger = _interopRequireDefault(require("@terrestris/base-util/dist/Logger"));

var _MapUtil = _interopRequireDefault(require("@terrestris/ol-util/dist/MapUtil/MapUtil"));

var _constants = require("../../constants");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Option = _select["default"].Option;

var _isInteger = require('lodash/isInteger');

var _isEmpty = require('lodash/isEmpty');

var _isEqual = require('lodash/isEqual');

var _isFunction = require('lodash/isFunction');

var _reverse = require('lodash/reverse');

var _clone = require('lodash/clone');

/**
 * Class representing a scale combo to choose map scale via a dropdown menu.
 *
 * @class The ScaleCombo
 * @extends React.Component
 */
var ScaleCombo =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ScaleCombo, _React$Component);

  _createClass(ScaleCombo, null, [{
    key: "getDerivedStateFromProps",

    /**
     * The className added to this component.
     * @private
     */

    /**
     * The default props
     */

    /**
     * Invoked after the component is instantiated as well as when it
     * receives new props. It should return an object to update state, or null
     * to indicate that the new props do not require any state updates.
     *
     * @param nextProps The next properties.
     * @param prevState The previous state.
     */
    value: function getDerivedStateFromProps(nextProps, prevState) {
      if (_isInteger(nextProps.zoomLevel) && !_isEqual(nextProps.zoomLevel, prevState.zoomLevel)) {
        return {
          zoomLevel: nextProps.zoomLevel
        };
      }

      if (_isFunction(nextProps.onZoomLevelSelect) && !_isEqual(nextProps.onZoomLevelSelect, prevState.onZoomLevelSelect)) {
        return {
          onZoomLevelSelect: nextProps.onZoomLevelSelect
        };
      }

      return null;
    }
    /**
     * Create a scale combo.
     * @constructs ScaleCombo
     */

  }]);

  function ScaleCombo(props) {
    var _this;

    _classCallCheck(this, ScaleCombo);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(ScaleCombo).call(this, props));
    /**
     * The default onZoomLevelSelect function sets the resolution of the passed
     * map according to the selected Scale.
     *
     * @param selectedScale The selectedScale.
     */

    _defineProperty(_assertThisInitialized(_this), "className", "".concat(_constants.CSS_PREFIX, "scalecombo"));

    _defineProperty(_assertThisInitialized(_this), "zoomListener", function (evt) {
      var zoom = evt.target.getView().getZoom();
      var roundZoom = Math.round(zoom);

      if (!roundZoom) {
        roundZoom = 0;
      }

      _this.setState({
        zoomLevel: roundZoom
      });
    });

    _defineProperty(_assertThisInitialized(_this), "pushScale", function (scales, resolution, view) {
      var scale = _MapUtil["default"].getScaleForResolution(resolution, view.getProjection().getUnits());

      var roundScale = _MapUtil["default"].roundScale(scale);

      if (scales.includes(roundScale)) {
        return;
      }

      scales.push(roundScale);
    });

    _defineProperty(_assertThisInitialized(_this), "getOptionsFromMap", function () {
      var _this$props = _this.props,
          map = _this$props.map,
          resolutionsFilter = _this$props.resolutionsFilter;

      if (!_isEmpty(_this.state.scales)) {
        _Logger["default"].debug('Array with scales found. Returning');

        return [];
      }

      if (!map) {
        _Logger["default"].warn('Map component not found. Could not initialize options array.');

        return [];
      }

      var scales = [];
      var view = map.getView(); // use existing resolutions array if exists

      var resolutions = view.getResolutions();

      if (_isEmpty(resolutions)) {
        for (var currentZoomLevel = view.getMaxZoom(); currentZoomLevel >= view.getMinZoom(); currentZoomLevel--) {
          var resolution = view.getResolutionForZoom(currentZoomLevel);

          if (resolutionsFilter(resolution)) {
            _this.pushScale(scales, resolution, view);
          }
        }
      } else {
        var reversedResolutions = _reverse(_clone(resolutions));

        reversedResolutions.filter(resolutionsFilter).forEach(function (resolution) {
          _this.pushScale(scales, resolution, view);
        });
      }

      return scales;
    });

    _defineProperty(_assertThisInitialized(_this), "determineOptionKeyForZoomLevel", function (zoom) {
      if (!_isInteger(zoom) || _this.state.scales.length - 1 - zoom < 0) {
        return undefined;
      }

      return _this.state.scales[_this.state.scales.length - 1 - zoom].toString();
    });

    var defaultOnZoomLevelSelect = function defaultOnZoomLevelSelect(selectedScale) {
      var mapView = props.map.getView();

      var calculatedResolution = _MapUtil["default"].getResolutionForScale(selectedScale, mapView.getProjection().getUnits());

      mapView.setResolution(calculatedResolution);
    };

    _this.state = {
      zoomLevel: props.zoomLevel || props.map.getView().getZoom(),
      onZoomLevelSelect: props.onZoomLevelSelect || defaultOnZoomLevelSelect,
      scales: props.scales || _this.getOptionsFromMap()
    };

    if (props.syncWithMap) {
      props.map.on('moveend', _this.zoomListener);
    }

    return _this;
  }
  /**
   * Invoked immediately after updating occurs. This method is not called for
   * the initial render.
   *
   * @param prevProps The previous props.
   */


  _createClass(ScaleCombo, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var _this$props2 = this.props,
          map = _this$props2.map,
          syncWithMap = _this$props2.syncWithMap;

      if (!_isEqual(syncWithMap, prevProps.syncWithMap)) {
        if (syncWithMap) {
          map.on('moveend', this.zoomListener);
        } else {
          map.un('moveend', this.zoomListener);
        }
      }
    }
    /**
     * Set the zoomLevel of the to the ScaleCombo.
     *
     * @param evt The 'moveend' event
     * @private
     */

  }, {
    key: "render",

    /**
     * The render function.
     */
    value: function render() {
      var _this$props3 = this.props,
          style = _this$props3.style,
          className = _this$props3.className;
      var _this$state = this.state,
          onZoomLevelSelect = _this$state.onZoomLevelSelect,
          scales = _this$state.scales,
          zoomLevel = _this$state.zoomLevel;
      var finalClassName = className ? "".concat(className, " ").concat(this.className) : this.className;
      var options = scales.map(function (roundScale) {
        return React.createElement(Option, {
          key: roundScale,
          value: roundScale.toString()
        }, "1:".concat(roundScale.toLocaleString()));
      });
      return React.createElement(_select["default"], {
        showSearch: true,
        onChange: onZoomLevelSelect,
        filterOption: function filterOption(input, option) {
          return option.key.toString().startsWith(input);
        },
        value: this.determineOptionKeyForZoomLevel(zoomLevel),
        size: "small",
        style: style,
        className: finalClassName
      }, options);
    }
  }]);

  return ScaleCombo;
}(React.Component);

_defineProperty(ScaleCombo, "defaultProps", {
  resolutionsFilter: function resolutionsFilter() {
    return true;
  },
  style: {
    width: 100
  },
  scales: [],
  syncWithMap: true
});

var _default = ScaleCombo;
exports["default"] = _default;
