// @terrestris/react-geo/dist/Button/DigitizeButton/DigitizeButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

require("antd/es/modal/style");

var _modal = _interopRequireDefault(require("antd/es/modal"));

require("antd/es/input/style");

var _input = _interopRequireDefault(require("antd/es/input"));

var React = _interopRequireWildcard(require("react"));

var _Vector = _interopRequireDefault(require("ol/layer/Vector"));

var _Vector2 = _interopRequireDefault(require("ol/source/Vector"));

var _Collection = _interopRequireDefault(require("ol/Collection"));

var _Style = _interopRequireDefault(require("ol/style/Style"));

var _Stroke = _interopRequireDefault(require("ol/style/Stroke"));

var _Fill = _interopRequireDefault(require("ol/style/Fill"));

var _Circle = _interopRequireDefault(require("ol/style/Circle"));

var _Text = _interopRequireDefault(require("ol/style/Text"));

var _Draw = _interopRequireWildcard(require("ol/interaction/Draw"));

var _Select = _interopRequireDefault(require("ol/interaction/Select"));

var _Modify = _interopRequireDefault(require("ol/interaction/Modify"));

var _Translate = _interopRequireDefault(require("ol/interaction/Translate"));

var _condition = require("ol/events/condition");

var _ToggleButton = _interopRequireDefault(require("../ToggleButton/ToggleButton"));

var _MapUtil = _interopRequireDefault(require("@terrestris/ol-util/dist/MapUtil/MapUtil"));

var _StringUtil = _interopRequireDefault(require("@terrestris/base-util/dist/StringUtil/StringUtil"));

var _AnimateUtil = _interopRequireDefault(require("@terrestris/ol-util/dist/AnimateUtil/AnimateUtil"));

var _Logger = _interopRequireDefault(require("@terrestris/base-util/dist/Logger"));

var _constants = require("../../constants");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var TextArea = _input["default"].TextArea;

var _isFunction = require('lodash/isFunction');

/**
 * The DigitizeButton.
 *
 * @class The DigitizeButton
 * @extends React.Component
 *
 */
var DigitizeButton =
/*#__PURE__*/
function (_React$Component) {
  _inherits(DigitizeButton, _React$Component);

  /**
   * The className added to this component.
   *
   * @private
   */

  /**
   * Currently existing digitize features as collection.
   *
   * @private
   */

  /**
   * The layer used for the digitization.
   *
   * @private
   */

  /**
   * Currently drawn feature which should be represent as label or postit.
   * @private
   */

  /**
   * Reference to OL select interaction which will be used in edit mode.
   * @private
   */

  /**
   * The interactions.
   * @private
   */

  /**
   * Name of point draw type.
   * @private
   */

  /**
   * Name of line string draw type.
   * @private
   */

  /**
   * Name of polygon draw type.
   * @private
   */

  /**
   * Name of circle draw type.
   * @private
   */

  /**
   * Name of rectangle draw type.
   * @private
   */

  /**
   * Name of text draw type.
   * @private
   */

  /**
   * Name of copy edit type.
   * @private
   */

  /**
   * Name of edit edit type.
   * @private
   */

  /**
   * Name of delete edit type.
   * @private
   */

  /**
   * Default fill color used in style object of drawn features.
   */

  /**
   * Default stroke color used in style object of drawn features.
   */

  /**
   * Hit detection in pixels used for select interaction.
   */

  /**
   * Default style for digitized points.
   */

  /**
   * Default style for digitized lines.
   */

  /**
   * Default style for digitized polygons or circles.
   */

  /**
   * Default style for digitized labels.
   */

  /**
   * The default properties.
   */

  /**
   * Creates the DigitizeButton.
   *
   * @constructs DigitizeButton
   */
  function DigitizeButton(props) {
    var _this;

    _classCallCheck(this, DigitizeButton);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(DigitizeButton).call(this, props));

    _defineProperty(_assertThisInitialized(_this), "className", "".concat(_constants.CSS_PREFIX, "digitizebutton"));

    _defineProperty(_assertThisInitialized(_this), "_digitizeFeatures", null);

    _defineProperty(_assertThisInitialized(_this), "_digitizeLayer", null);

    _defineProperty(_assertThisInitialized(_this), "_digitizeTextFeature", null);

    _defineProperty(_assertThisInitialized(_this), "_selectInteraction", null);

    _defineProperty(_assertThisInitialized(_this), "_interactions", []);

    _defineProperty(_assertThisInitialized(_this), "onToggle", function (pressed) {
      var _this$props = _this.props,
          map = _this$props.map,
          drawType = _this$props.drawType,
          editType = _this$props.editType,
          drawStyle = _this$props.drawStyle,
          onFeatureSelect = _this$props.onFeatureSelect;

      _this.props.onToggle(pressed);

      if (_this._digitizeLayer) {
        _this._digitizeFeatures = _this._digitizeLayer.getSource().getFeaturesCollection();
      }

      if (pressed) {
        if (drawStyle) {
          _this._digitizeLayer.setStyle(_this.getDigitizeStyleFunction);
        }

        if (drawType) {
          _this.createDrawInteraction(pressed);
        } else if (editType) {
          _this.createSelectOrModifyInteraction();
        }
      } else {
        _this._interactions.forEach(function (i) {
          return map.removeInteraction(i);
        });

        if (drawType === DigitizeButton.TEXT_DRAW_TYPE) {
          _this._digitizeFeatures.un('add', _this.handleTextAdding);
        } else {
          if (_this._selectInteraction) {
            _this._selectInteraction.getFeatures().clear();
          }

          if (editType === DigitizeButton.DELETE_EDIT_TYPE) {
            _this._selectInteraction.un('select', _this.onFeatureRemove);
          }

          if (editType === DigitizeButton.COPY_EDIT_TYPE) {
            _this._selectInteraction.un('select', _this.onFeatureCopy);
          }

          if (_isFunction(onFeatureSelect) && editType === DigitizeButton.EDIT_EDIT_TYPE) {
            _this._selectInteraction.un('select', onFeatureSelect);
          }

          map.un('pointermove', _this.onPointerMove);
        }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "createDigitizeLayer", function () {
      var _this$props2 = _this.props,
          digitizeLayerName = _this$props2.digitizeLayerName,
          map = _this$props2.map;

      var digitizeLayer = _MapUtil["default"].getLayerByName(map, digitizeLayerName);

      if (!digitizeLayer) {
        digitizeLayer = new _Vector["default"]({
          name: digitizeLayerName,
          source: new _Vector2["default"]({
            features: new _Collection["default"]()
          })
        });
        map.addLayer(digitizeLayer);
      }

      digitizeLayer.setStyle(_this.getDigitizeStyleFunction);
      _this._digitizeLayer = digitizeLayer;
    });

    _defineProperty(_assertThisInitialized(_this), "getDigitizeStyleFunction", function (feature) {
      var drawStyle = _this.props.drawStyle;
      var styleObj;

      if (!feature.getGeometry()) {
        return;
      }

      switch (feature.getGeometry().getType()) {
        case DigitizeButton.POINT_DRAW_TYPE:
          {
            if (!feature.get('isLabel')) {
              styleObj = drawStyle || new _Style["default"]({
                image: new _Circle["default"]({
                  radius: 7,
                  fill: new _Fill["default"]({
                    color: DigitizeButton.DEFAULT_FILL_COLOR
                  }),
                  stroke: new _Stroke["default"]({
                    color: DigitizeButton.DEFAULT_STROKE_COLOR
                  })
                })
              });
            } else {
              styleObj = drawStyle || new _Style["default"]({
                text: new _Text["default"]({
                  text: feature.get('label'),
                  offsetX: 5,
                  offsetY: 5,
                  font: '12px sans-serif',
                  fill: new _Fill["default"]({
                    color: DigitizeButton.DEFAULT_FILL_COLOR
                  }),
                  stroke: new _Stroke["default"]({
                    color: DigitizeButton.DEFAULT_STROKE_COLOR
                  })
                })
              });
            }

            return styleObj;
          }

        case DigitizeButton.LINESTRING_DRAW_TYPE:
          {
            styleObj = drawStyle || new _Style["default"]({
              stroke: new _Stroke["default"]({
                color: DigitizeButton.DEFAULT_STROKE_COLOR,
                width: 2
              })
            });
            return styleObj;
          }

        case DigitizeButton.POLYGON_DRAW_TYPE:
        case DigitizeButton.CIRCLE_DRAW_TYPE:
          {
            styleObj = drawStyle || new _Style["default"]({
              fill: new _Fill["default"]({
                color: DigitizeButton.DEFAULT_FILL_COLOR
              }),
              stroke: new _Stroke["default"]({
                color: DigitizeButton.DEFAULT_STROKE_COLOR,
                width: 2
              })
            });
            return styleObj;
          }

        default:
          {
            break;
          }
      }
    });

    _defineProperty(_assertThisInitialized(_this), "getSelectedStyleFunction", function (feature, res, text) {
      var _this$props3 = _this.props,
          selectFillColor = _this$props3.selectFillColor,
          selectStrokeColor = _this$props3.selectStrokeColor;

      if (feature.get('label')) {
        text = feature.get('label');
      }

      return new _Style["default"]({
        image: new _Circle["default"]({
          radius: 7,
          fill: new _Fill["default"]({
            color: selectFillColor
          }),
          stroke: new _Stroke["default"]({
            color: selectStrokeColor
          })
        }),
        text: new _Text["default"]({
          text: text ? text : '',
          offsetX: 5,
          offsetY: 5,
          font: '12px sans-serif',
          fill: new _Fill["default"]({
            color: selectFillColor
          }),
          stroke: new _Stroke["default"]({
            color: selectStrokeColor
          })
        }),
        stroke: new _Stroke["default"]({
          color: selectStrokeColor,
          width: 2
        }),
        fill: new _Fill["default"]({
          color: selectFillColor
        })
      });
    });

    _defineProperty(_assertThisInitialized(_this), "createDrawInteraction", function (pressed) {
      var _this$props4 = _this.props,
          drawType = _this$props4.drawType,
          map = _this$props4.map,
          onDrawEnd = _this$props4.onDrawEnd,
          onDrawStart = _this$props4.onDrawStart,
          digitizeLayerName = _this$props4.digitizeLayerName,
          drawInteractionConfig = _this$props4.drawInteractionConfig;
      var geometryFunction;
      var type = drawType; // check whether the digitizeLayer is in map and set it from state if not

      var digitizeLayer = _MapUtil["default"].getLayerByName(map, digitizeLayerName);

      if (!digitizeLayer) {
        map.addLayer(_this._digitizeLayer);
      }

      if (drawType === DigitizeButton.RECTANGLE_DRAW_TYPE) {
        geometryFunction = (0, _Draw.createBox)();
        type = DigitizeButton.CIRCLE_DRAW_TYPE;
      } else if (drawType === DigitizeButton.TEXT_DRAW_TYPE) {
        type = DigitizeButton.POINT_DRAW_TYPE;

        _this._digitizeFeatures.on('add', _this.handleTextAdding);
      }

      var drawInteraction = new _Draw["default"](_objectSpread({
        source: _this._digitizeLayer.getSource(),
        type: type,
        geometryFunction: geometryFunction,
        style: _this.getDigitizeStyleFunction,
        freehandCondition: _condition.never
      }, drawInteractionConfig));

      if (onDrawEnd) {
        drawInteraction.on('drawend', onDrawEnd);
      }

      if (onDrawStart) {
        drawInteraction.on('drawstart', onDrawStart);
      }

      map.addInteraction(drawInteraction);
      _this._interactions = [drawInteraction];
      drawInteraction.setActive(pressed);
    });

    _defineProperty(_assertThisInitialized(_this), "createSelectOrModifyInteraction", function () {
      var _this$props5 = _this.props,
          editType = _this$props5.editType,
          map = _this$props5.map,
          selectInteractionConfig = _this$props5.selectInteractionConfig,
          modifyInteractionConfig = _this$props5.modifyInteractionConfig,
          translateInteractionConfig = _this$props5.translateInteractionConfig,
          onFeatureSelect = _this$props5.onFeatureSelect;
      _this._selectInteraction = new _Select["default"](_objectSpread({
        condition: _condition.singleClick,
        hitTolerance: DigitizeButton.HIT_TOLERANCE,
        style: _this.getSelectedStyleFunction
      }, selectInteractionConfig));

      if (editType === DigitizeButton.DELETE_EDIT_TYPE) {
        _this._selectInteraction.on('select', _this.onFeatureRemove);
      } else if (editType === DigitizeButton.COPY_EDIT_TYPE) {
        _this._selectInteraction.on('select', _this.onFeatureCopy);
      }

      if (_isFunction(onFeatureSelect) && editType === DigitizeButton.EDIT_EDIT_TYPE) {
        _this._selectInteraction.on('select', onFeatureSelect);
      }

      var interactions = [_this._selectInteraction];

      if (editType === DigitizeButton.EDIT_EDIT_TYPE) {
        var edit = new _Modify["default"](_objectSpread({
          features: _this._selectInteraction.getFeatures(),
          deleteCondition: _condition.singleClick,
          style: _this.getSelectedStyleFunction
        }, modifyInteractionConfig));
        edit.on('modifystart', _this.onModifyStart);
        edit.on('modifyend', _this.onModifyEnd);
        var translate = new _Translate["default"](_objectSpread({
          features: _this._selectInteraction.getFeatures()
        }, translateInteractionConfig));
        translate.on('translatestart', _this.onTranslateStart);
        translate.on('translateend', _this.onTranslateEnd);
        translate.on('translating', _this.onTranslating);
        interactions.push(edit, translate);
      }

      interactions.forEach(function (i) {
        return map.addInteraction(i);
      });
      map.on('pointermove', _this.onPointerMove);
      _this._interactions = interactions;
    });

    _defineProperty(_assertThisInitialized(_this), "onFeatureRemove", function (evt) {
      var _this$props6 = _this.props,
          onFeatureRemove = _this$props6.onFeatureRemove,
          onFeatureSelect = _this$props6.onFeatureSelect;

      if (_isFunction(onFeatureRemove)) {
        onFeatureRemove(evt);
      }

      if (_isFunction(onFeatureSelect)) {
        onFeatureSelect(evt);
      }

      var feat = evt.selected[0];

      _this._selectInteraction.getFeatures().remove(feat);

      _this._digitizeLayer.getSource().removeFeature(feat);

      _this.props.map.renderSync();
    });

    _defineProperty(_assertThisInitialized(_this), "onFeatureCopy", function (evt) {
      var _this$props7 = _this.props,
          map = _this$props7.map,
          onFeatureCopy = _this$props7.onFeatureCopy,
          onFeatureSelect = _this$props7.onFeatureSelect;
      var feat = evt.selected[0];

      if (!feat) {
        return;
      }

      if (_isFunction(onFeatureCopy)) {
        onFeatureCopy(evt);
      }

      if (_isFunction(onFeatureSelect)) {
        onFeatureSelect(evt);
      }

      var copy = feat.clone();
      copy.setStyle(_this.getDigitizeStyleFunction(feat));

      _this._digitizeFeatures.push(copy);

      _AnimateUtil["default"].moveFeature(map, _this._digitizeLayer, copy, 500, 50, _this.getDigitizeStyleFunction(feat));
    });

    _defineProperty(_assertThisInitialized(_this), "onModifyStart", function (evt) {
      var onModifyStart = _this.props.onModifyStart;

      if (_isFunction(onModifyStart)) {
        onModifyStart(evt);
      }

      var feature = evt.features.getArray()[0];

      if (feature.get('isLabel')) {
        _this._digitizeTextFeature = feature;
        var textLabel = '';

        var featureStyle = _this.getDigitizeStyleFunction(feature);

        if (featureStyle && featureStyle.getText()) {
          textLabel = featureStyle.getText().getText();
        } else if (feature.get('label')) {
          textLabel = feature.get('label');
        }

        _this.setState({
          showLabelPrompt: true,
          textLabel: textLabel
        });
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onModifyEnd", function (evt) {
      var onModifyEnd = _this.props.onModifyEnd;

      if (_isFunction(onModifyEnd)) {
        onModifyEnd(evt);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onTranslateStart", function (evt) {
      var onTranslateStart = _this.props.onTranslateStart;

      if (_isFunction(onTranslateStart)) {
        onTranslateStart(evt);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onTranslateEnd", function (evt) {
      var onTranslateEnd = _this.props.onTranslateEnd;

      if (_isFunction(onTranslateEnd)) {
        onTranslateEnd(evt);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "onTranslating", function (evt) {
      var onTranslating = _this.props.onTranslating;

      if (_isFunction(onTranslating)) {
        onTranslating(evt);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleTextAdding", function (evt) {
      _this.setState({
        showLabelPrompt: true
      });

      _this._digitizeTextFeature = evt.element;

      _this._digitizeTextFeature.set('isLabel', true);
    });

    _defineProperty(_assertThisInitialized(_this), "onModalLabelOk", function () {
      var onModalLabelOk = _this.props.onModalLabelOk;

      _this.setState({
        showLabelPrompt: false
      }, function () {
        _this.setTextOnFeature(_this._digitizeTextFeature, onModalLabelOk);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onModalLabelCancel", function (event) {
      var onModalLabelCancel = _this.props.onModalLabelCancel;

      _this.setState({
        showLabelPrompt: false,
        textLabel: ''
      }, function () {
        if (!(_this._interactions.length > 1)) {
          _this._digitizeFeatures.remove(_this._digitizeTextFeature);

          _this._digitizeTextFeature = null;
        }

        if (_isFunction(onModalLabelCancel)) {
          onModalLabelCancel(event);
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "setTextOnFeature", function (feat, onModalOkCbk) {
      var maxLabelLineLength = _this.props.maxLabelLineLength;
      var textLabel = _this.state.textLabel;
      var label = textLabel;

      if (maxLabelLineLength) {
        label = _StringUtil["default"].stringDivider(textLabel, maxLabelLineLength, '\n');
      }

      feat.set('label', label);

      _this.setState({
        textLabel: ''
      }, function () {
        if (_isFunction(onModalOkCbk)) {
          onModalOkCbk(feat, label);
        }
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onLabelChange", function (evt) {
      _this.setState({
        textLabel: evt.target.value
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onPointerMove", function (evt) {
      if (evt.dragging) {
        return;
      }

      var _this$props8 = _this.props,
          map = _this$props8.map,
          digitizeLayerName = _this$props8.digitizeLayerName;
      var pixel = map.getEventPixel(evt.originalEvent);
      var hit = map.hasFeatureAtPixel(pixel, {
        layerFilter: function layerFilter(l) {
          return l.get('name') === digitizeLayerName;
        },
        hitTolerance: DigitizeButton.HIT_TOLERANCE
      });
      map.getTargetElement().style.cursor = hit ? 'pointer' : '';
    });

    if (!props.drawType && !props.editType) {
      _Logger["default"].warn('Neither "drawType" nor "editType" was provided. Digitize ' + 'button won\'t work properly!');
    }

    _this.state = {
      showLabelPrompt: false,
      textLabel: ''
    };
    return _this;
  }
  /**
   * `componentDidMount` method of the DigitizeButton. Just calls
   * `createDigitizeLayer` method.
   */


  _createClass(DigitizeButton, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.createDigitizeLayer();
    }
    /**
     * Called on componentWillUnmount lifecycle.
     */

  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      var map = this.props.map;

      this._interactions.forEach(function (i) {
        return map.removeInteraction(i);
      });
    }
    /**
     * Called when the digitize button is toggled. If the button state is pressed,
     * the appropriate draw, modify or select interaction will be created.
     * Otherwise, by untoggling, the same previously created interaction will be
     * removed from the map.
     *
     * @param pressed Whether the digitize button is pressed or not.
     */

  }, {
    key: "render",

    /**
     * The render function.
     */
    value: function render() {
      var _this$props9 = this.props,
          className = _this$props9.className,
          map = _this$props9.map,
          drawType = _this$props9.drawType,
          editType = _this$props9.editType,
          digitizeLayerName = _this$props9.digitizeLayerName,
          drawStyle = _this$props9.drawStyle,
          selectFillColor = _this$props9.selectFillColor,
          selectStrokeColor = _this$props9.selectStrokeColor,
          modalPromptTitle = _this$props9.modalPromptTitle,
          modalPromptOkButtonText = _this$props9.modalPromptOkButtonText,
          modalPromptCancelButtonText = _this$props9.modalPromptCancelButtonText,
          onDrawStart = _this$props9.onDrawStart,
          onDrawEnd = _this$props9.onDrawEnd,
          onModifyStart = _this$props9.onModifyStart,
          onModifyEnd = _this$props9.onModifyEnd,
          onTranslateStart = _this$props9.onTranslateStart,
          onTranslateEnd = _this$props9.onTranslateEnd,
          onTranslating = _this$props9.onTranslating,
          onFeatureRemove = _this$props9.onFeatureRemove,
          onFeatureCopy = _this$props9.onFeatureCopy,
          onFeatureSelect = _this$props9.onFeatureSelect,
          drawInteractionConfig = _this$props9.drawInteractionConfig,
          selectInteractionConfig = _this$props9.selectInteractionConfig,
          modifyInteractionConfig = _this$props9.modifyInteractionConfig,
          translateInteractionConfig = _this$props9.translateInteractionConfig,
          onToggle = _this$props9.onToggle,
          onModalLabelOk = _this$props9.onModalLabelOk,
          onModalLabelCancel = _this$props9.onModalLabelCancel,
          maxLabelLineLength = _this$props9.maxLabelLineLength,
          passThroughProps = _objectWithoutProperties(_this$props9, ["className", "map", "drawType", "editType", "digitizeLayerName", "drawStyle", "selectFillColor", "selectStrokeColor", "modalPromptTitle", "modalPromptOkButtonText", "modalPromptCancelButtonText", "onDrawStart", "onDrawEnd", "onModifyStart", "onModifyEnd", "onTranslateStart", "onTranslateEnd", "onTranslating", "onFeatureRemove", "onFeatureCopy", "onFeatureSelect", "drawInteractionConfig", "selectInteractionConfig", "modifyInteractionConfig", "translateInteractionConfig", "onToggle", "onModalLabelOk", "onModalLabelCancel", "maxLabelLineLength"]);

      var finalClassName = className ? "".concat(className, " ").concat(this.className) : this.className;
      var btnWrapperClass = "".concat(_constants.CSS_PREFIX, "digitize-button-wrapper");
      return React.createElement("span", {
        className: btnWrapperClass
      }, React.createElement(_ToggleButton["default"], _extends({
        onToggle: this.onToggle,
        className: finalClassName
      }, passThroughProps)), this.state.showLabelPrompt ? React.createElement(_modal["default"], {
        title: modalPromptTitle,
        okText: modalPromptOkButtonText,
        cancelText: modalPromptCancelButtonText,
        visible: this.state.showLabelPrompt,
        closable: false,
        onOk: this.onModalLabelOk,
        onCancel: this.onModalLabelCancel
      }, React.createElement(TextArea, {
        value: this.state.textLabel,
        onChange: this.onLabelChange,
        autosize: true
      })) : null);
    }
  }]);

  return DigitizeButton;
}(React.Component);

_defineProperty(DigitizeButton, "POINT_DRAW_TYPE", 'Point');

_defineProperty(DigitizeButton, "LINESTRING_DRAW_TYPE", 'LineString');

_defineProperty(DigitizeButton, "POLYGON_DRAW_TYPE", 'Polygon');

_defineProperty(DigitizeButton, "CIRCLE_DRAW_TYPE", 'Circle');

_defineProperty(DigitizeButton, "RECTANGLE_DRAW_TYPE", 'Rectangle');

_defineProperty(DigitizeButton, "TEXT_DRAW_TYPE", 'Text');

_defineProperty(DigitizeButton, "COPY_EDIT_TYPE", 'Copy');

_defineProperty(DigitizeButton, "EDIT_EDIT_TYPE", 'Edit');

_defineProperty(DigitizeButton, "DELETE_EDIT_TYPE", 'Delete');

_defineProperty(DigitizeButton, "DEFAULT_FILL_COLOR", 'rgba(154, 26, 56, 0.5)');

_defineProperty(DigitizeButton, "DEFAULT_STROKE_COLOR", 'rgba(154, 26, 56, 0.8)');

_defineProperty(DigitizeButton, "HIT_TOLERANCE", 5);

_defineProperty(DigitizeButton, "DEFAULT_POINT_STYLE", new _Style["default"]({
  image: new _Circle["default"]({
    radius: 7,
    fill: new _Fill["default"]({
      color: DigitizeButton.DEFAULT_FILL_COLOR
    }),
    stroke: new _Stroke["default"]({
      color: DigitizeButton.DEFAULT_STROKE_COLOR
    })
  })
}));

_defineProperty(DigitizeButton, "DEFAULT_LINESTRING_STYLE", new _Style["default"]({
  stroke: new _Stroke["default"]({
    color: DigitizeButton.DEFAULT_STROKE_COLOR,
    width: 2
  })
}));

_defineProperty(DigitizeButton, "DEFAULT_POLYGON_STYLE", new _Style["default"]({
  fill: new _Fill["default"]({
    color: DigitizeButton.DEFAULT_FILL_COLOR
  }),
  stroke: new _Stroke["default"]({
    color: DigitizeButton.DEFAULT_STROKE_COLOR,
    width: 2
  })
}));

_defineProperty(DigitizeButton, "DEFAULT_TEXT_STYLE", new _Style["default"]({
  text: new _Text["default"]({
    text: '',
    offsetX: 5,
    offsetY: 5,
    font: '12px sans-serif',
    fill: new _Fill["default"]({
      color: DigitizeButton.DEFAULT_FILL_COLOR
    }),
    stroke: new _Stroke["default"]({
      color: DigitizeButton.DEFAULT_STROKE_COLOR
    })
  })
}));

_defineProperty(DigitizeButton, "defaultProps", {
  digitizeLayerName: 'react-geo_digitize',
  selectFillColor: 'rgba(240, 240, 90, 0.5)',
  selectStrokeColor: 'rgba(220, 120, 20, 0.8)',
  modalPromptTitle: 'Label',
  modalPromptOkButtonText: 'Ok',
  modalPromptCancelButtonText: 'Cancel',
  drawInteractionConfig: {},
  selectInteractionConfig: {},
  modifyInteractionConfig: {},
  translateInteractionConfig: {},
  onToggle: function onToggle() {
    return undefined;
  }
});

var _default = DigitizeButton;
exports["default"] = _default;
